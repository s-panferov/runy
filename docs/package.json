{
  "name": "@runy-dev/website",
  "type": "module",
  "version": "0.0.1",
  "scripts": {
    "dev": "astro dev",
    "start": "astro dev",
    "build": "astro build",
    "preview": "astro preview",
    "astro": "astro"
  },
  "dependencies": {
    "@astrojs/react": "^4.3.0",
    "@astrojs/starlight": "^0.35.1",
    "@astrojs/starlight-tailwind": "^4.0.1",
    "@fontsource/geist-mono": "^5.2.6",
    "@fontsource/geist-sans": "^5.2.5",
    "@iconify-json/mdi": "^1.2.3",
    "@stylexjs/stylex": "^0.14.2",
    "@tailwindcss/vite": "^4.1.11",
    "@types/react": "^19.1.8",
    "@types/react-dom": "^19.1.6",
    "astro": "^5.6.1",
    "astro-d2": "^0.8.0",
    "astro-icon": "^1.1.5",
    "react": "^19.1.0",
    "react-dom": "^19.1.0",
    "sharp": "^0.34.2",
    "starlight-theme-nova": "^0.9.0",
    "tailwindcss": "^4.1.11",
    "vite-plugin-stylex": "^0.13.0"
  }
}
