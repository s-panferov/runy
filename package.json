{
  "name": "@runy-dev/monorepo",
  "type": "module",
  "version": "0.0.1",
  "license": "MIT",
  "private": true,
  "scripts": {
    "test": "NODE_OPTIONS=\"$NODE_OPTIONS --experimental-vm-modules --no-warnings\" jest",
    "check": "tsc -b"
  },
  "dependencies": {
    "@microsoft/api-extractor": "^7.52.8",
    "@rollup/plugin-commonjs": "^28.0.6",
    "@rollup/plugin-json": "^6.1.0",
    "@rollup/plugin-swc": "^0.4.0",
    "@rollup/plugin-node-resolve": "^16.0.1",
    "@rsbuild/core": "^1.4.8",
    "@swc/helpers": "^0.5.17",
    "@types/node": "^24.0.15",
    "rollup": "^4.45.1",
    "typescript": "5.8.3",
    "vitest": "^3.2.4"
  },
  "devDependencies": {
    "@semantic-release/commit-analyzer": "^13.0.1",
    "@semantic-release/exec": "^7.1.0",
    "@semantic-release/github": "^11.0.3",
    "@semantic-release/release-notes-generator": "^14.0.3",
    "semantic-release": "^24.2.7",
    "semantic-release-pnpm": "^1.0.2"
  }
}
