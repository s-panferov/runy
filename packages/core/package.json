{
  "name": "@runy-dev/core",
  "version": "0.0.4",
  "private": false,
  "dependencies": {
    "@runy-dev/proto": "workspace:*",
    "@types/node": "^24.0.15",
    "execa": "^9.6.0",
    "winston": "^3.17.0"
  },
  "exports": {
    ".": {
      "import": "./src/index.ts"
    }
  }
}
