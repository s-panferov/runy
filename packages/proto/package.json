{
  "name": "@runy-dev/proto",
  "version": "0.0.4",
  "private": false,
  "scripts": {
    "generate": "sh ./gen.sh"
  },
  "devDependencies": {
    "ts-proto": "^2.7.5",
    "@bufbuild/protobuf": "^2.6.1",
    "@grpc/grpc-js": "^1.13.4",
    "grpc-web": "^1.5.0",
    "protoc-gen-js": "^3.21.4-4",
    "@types/google-protobuf": "^3.15.12",
    "google-protobuf": "^3.21.4",
    "rxjs": "^7.8.2"
  },
  "exports": {
    "./rpc": {
      "default": "./src/rpc.ts"
    },
    "./browser": {
      "default": "./src/browser.ts"
    },
    "./process": {
      "default": "./src/process.ts"
    },
    ".": {
      "default": "./src/index.ts"
    }
  }
}
