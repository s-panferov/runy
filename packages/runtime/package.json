{
  "name": "@runy-build/runtime",
  "version": "0.1.0",
  "type": "module",
  "license": "MIT",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "dependencies": {
    "@runy-build/schema": "workspace:*",
    "execa": "^9.5.3",
    "vscode-jsonrpc": "^8.2.1",
    "winston": "^3.17.0"
  },
  "exports": {
    ".": "./dist/index.js",
    "./BUILD": "./BUILD.ts"
  },
  "files": [
    "dist/index.js",
    "dist/index.js.map",
    "dist/index.d.ts",
    "package.json",
    "README.md",
    "LICENSE"
  ]
}
