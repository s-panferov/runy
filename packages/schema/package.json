{
  "name": "@runy-build/schema",
  "version": "0.1.0",
  "type": "module",
  "license": "MIT",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "dependencies": {
    "@runy-build/types": "workspace:*",
    "@types/object-hash": "3.0.6",
    "object-hash": "*",
    "zod": "^3.25.3",
    "zod-to-json-schema": "^3.24.5"
  },
  "devDependencies": {
    "@types/node": "^22.15.19",
    "node-polyfill-webpack-plugin": "*"
  },
  "exports": {
    ".": "./dist/index.js",
    "./BUILD": "./BUILD.ts"
  },
  "files": [
    "dist/index.js",
    "dist/index.js.map",
    "dist/index.d.ts",
    "package.json",
    "README.md",
    "LICENSE"
  ]
}
