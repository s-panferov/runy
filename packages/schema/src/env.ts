import { dirname } from "node:path";
import process from "node:process";
import assert from "node:assert";

export let WORKSPACE_ROOT = "";

try {
  WORKSPACE_ROOT = process.env["BUILD_WORKSPACE"] || process.cwd();
} catch {
  // no access
}

export function setWorkspaceRoot(path: string, parent = false): void {
  assert(!!path);

  // Check if the path is a URL
  try {
    const url = new URL(path);
    // If it's a valid URL, convert it to a pathname and use its parent
    WORKSPACE_ROOT = url.pathname;
  } catch {
    // Use the parent path by default
    WORKSPACE_ROOT = path;
  }

  if (parent) {
    WORKSPACE_ROOT = dirname(WORKSPACE_ROOT);
  }
}

export function setTestRoot(path: string) {
  assert(!!path);
  setWorkspaceRoot(path, true);
}
