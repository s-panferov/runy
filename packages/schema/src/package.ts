import { flag, Flag, Target } from "./index";

import { BuildSpecFactory } from "./spec";
import { registry } from "./registry";
import { SpecT } from "./target";
import { WORKSPACE_ROOT } from "./env";

import path from "path";

export class Package {
  constructor(path: string) {
    this.path = path;
  }

  targets = new Map<string, Target.Any>();
  flags = new Map<string, Flag<symbol, any>>();
  path: string;

  #defaultTarget?: Target.Any;

  target<const T extends SpecT>(
    name: string,
    func: BuildSpecFactory<T>
  ): Target<T & { Actions: {} }> {
    return new Target(this, name, func as any);
  }

  flag<const K extends symbol, const T extends flag.FlagType<any>>(
    symbol: K,
    ty: T
  ): Flag<K, T> {
    const flag = new Flag<K, T>(this, symbol, ty);
    this.flags.set(symbol.description!, flag);
    return flag;
  }

  input<const T extends SpecT>(
    name: string,
    func: BuildSpecFactory<T>
  ): Target<T & { Actions: {} }> {
    return new Target(this, name, func as any);
  }

  get defaultTarget(): Target.Any | undefined {
    return this.#defaultTarget;
  }

  default<T extends Target<any>>(target: T) {
    if (this.#defaultTarget) {
      this.#defaultTarget.default = false;
    }

    target.default = true;
    this.#defaultTarget = target;
  }

  exportTarget(target: Target.Any) {
    const exists = this.targets.get(target.name);
    if (exists) {
      throw new Error(
        `Duplicated target: "${target.name}" in ${target.package}`
      );
    }

    this.targets.set(target.name, target);
  }

  exportFlag(flag: Flag<symbol, any>) {
    const name = flag.key.description || "<flag>";
    const exists = this.flags.get(name);
    if (exists) {
      throw new Error(`Duplicated flag: "${name}" in ${flag.package}`);
    }

    this.flags.set(name, flag);
  }
}

export function getPackage(meta: ImportMeta | URL | string): Package {
  if (typeof meta === "string") {
    return registry.definePackage(meta);
  }

  let urlStr = "url" in meta ? meta.url : meta;
  const url = new URL(urlStr);

  let pkg = "";

  const filePath = decodeURIComponent(url.pathname);

  if (WORKSPACE_ROOT) {
    pkg = path.relative(WORKSPACE_ROOT, filePath);
  } else {
    // If BUILD_ROOT is not set, use the absolute file path
    pkg = filePath;
  }

  if (pkg.startsWith("./")) {
    pkg = pkg.slice(2);
  }

  if (pkg == ".") {
    pkg = "";
  }

  return registry.definePackage(pkg);
}
