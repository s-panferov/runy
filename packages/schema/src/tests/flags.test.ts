import { getPackage } from "../package";
import { BuildContext, BuildSpec, Schema, flag } from "../index";
import { setTestRoot } from "../env";
import { expect, test } from "vitest";

test("flags", () => {
  setTestRoot(import.meta.url);

  const mod = getPackage(import.meta);
  const f1 = mod.flag(Symbol("f1"), flag.string());

  const a = mod.target("a", (spec: BuildSpec, ctx) => {
    spec.out({
      f1: f1.in(ctx),
    });

    return spec;
  });

  const b = mod.target("b", (spec: BuildSpec, ctx) => {
    spec.out({
      a: a.out(ctx.with(f1.is("override"))),
      fi: f1.in(ctx),
    });

    return spec;
  });

  const ctx = new BuildContext().with(f1.is("original"));
  const schema = Schema.convert(b.out(ctx));

  expect(schema).toMatchSnapshot();
});
