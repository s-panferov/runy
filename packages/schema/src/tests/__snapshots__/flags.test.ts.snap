// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flags 1`] = `
{
  "builds": {
    "Ldo4NR+irhxU0yjv81erUWNhhAo=": {
      "ctx": {
        "$ref": "rgxRjKqBunOoHkrR0YGgen1DsYw=",
        "kind": "context",
      },
      "spec": {
        "key": [],
        "out": {
          "a": {
            "$ref": "y3fNENeOP+cjAS45GnFnf1UBNfg=",
            "kind": "build",
            "phase": "out",
          },
          "fi": "original",
        },
      },
      "target": {
        "$ref": "lJ+Vvyx/l4y+WSA3cnd6QY3D1vU=",
        "kind": "target",
      },
    },
    "y3fNENeOP+cjAS45GnFnf1UBNfg=": {
      "ctx": {
        "$ref": "U1wwAaqukwfarhkuFGe2QSaTdXY=",
        "kind": "context",
      },
      "spec": {
        "key": [],
        "out": {
          "f1": "override",
        },
      },
      "target": {
        "$ref": "PKV16cZy/DeLaifTrfeD9Zkt+Js=",
        "kind": "target",
      },
    },
  },
  "contexts": {
    "U1wwAaqukwfarhkuFGe2QSaTdXY=": {
      "flags": {
        "0L2zKz2jncrRJgITr64wWtgx8AY=": "override",
      },
      "parent": undefined,
    },
    "rgxRjKqBunOoHkrR0YGgen1DsYw=": {
      "flags": {
        "0L2zKz2jncrRJgITr64wWtgx8AY=": "original",
      },
      "parent": undefined,
    },
  },
  "flags": {
    "0L2zKz2jncrRJgITr64wWtgx8AY=": {
      "key": "f1",
      "package": "flags.test.ts",
      "ty": {
        "kind": "string",
      },
    },
  },
  "root": {
    "$ref": "Ldo4NR+irhxU0yjv81erUWNhhAo=",
    "kind": "build",
    "phase": "out",
  },
  "targets": {
    "PKV16cZy/DeLaifTrfeD9Zkt+Js=": {
      "default": false,
      "name": "a",
      "package": "flags.test.ts",
    },
    "lJ+Vvyx/l4y+WSA3cnd6QY3D1vU=": {
      "default": false,
      "name": "b",
      "package": "flags.test.ts",
    },
  },
}
`;
